Archive manager

white-axe opened the pull request:
**Description**
This pull request adds an archive manager for creating and extracting rgssad, rgss2a and rgss3a archives. It can be found in a new "Tools" menu in the top bar. It tries to load from the current project by default, but the user can choose a different location for the archive or source directory and there doesn't have to be a project loaded to use the archive manager.

The archive manager has a filesystem view for viewing the contents of archives or the contents of source directories for archives. It supports basic multi-file/directory selection with the Control and Shift keys, and is lazy (does not read the contents of directories until the user expands the tree for that directory).

I made some changes to the filesystems as well:

- The `create_dir` function in the filesystems is now recursive, that is it now creates missing parent directories instead of erroring, and it now does nothing if the destination directory already exists instead of erroring. I needed this for the archive manager, and I didn't see the original behaviour being used anywhere anyways. In fact, the new behaviour is easier to implement in the web and archiver filesystems than the old one, and prevents us from having to check if the .luminol directory exists before creating it.
- The native and web file implementations now have `from_file_picker` for choosing a single file using a file picker.
- The native and web file implementations now have a `save` method for showing a save file dialog.
- The web filesystem no longer stores directory handles to IndexedDB by default, rather they will only be stored if `.save_to_idb()` is called, which consumes the directory handle. This helps prevent stray handles from building up in IndexedDB and also increases the ergonomics of the web filesystem since we don't need to store directory handles to IndexedDB most of the time.
- The file trait is now automatically implemented for mutable references to anything that also implements the file trait. This allows, for example, the archiver filesystem to borrow archive files instead of taking ownership of them.
- The filesystem and file traits no longer require static lifetimes because that wasn't required before by anything other than the audio handler.
- The files of the native and web filesystems now also implement `AsyncRead + AsyncWrite + AsyncSeek + Unpin`.

**Testing**
Use the archive manager to extract an archive created by RPG Maker, and try loading an archive created by the archive manager into the archive manager again or mkxp-z, etc.

**Checklist**

- [x] Run `cargo fmt`.
- [x] Run `cargo clippy`. If applicable, add:
  - [x] `--target wasm32-unknown-unknown`
- [x] Run `cargo build --release` 
- [x] If applicable, run `trunk build --release`

white-axe committed:
feat: start creating the UI for the archive manager

white-axe committed:
refactor(filesystem): add `luminol_core::spawn_future` for cross-platform spawning

white-axe committed:
chore(filesystem): fix redundant closures

white-axe committed:
refactor(project manager): move `frame` into `update_state`

white-axe committed:
refactor(filesystem): don't store in IndexedDB unless requested

In web builds, we usually don't need to store folder handles in
IndexedDB, so the old system of storing folder handles in IndexedDB by
default and deleting them from IndexedDB if requested was clunky.

This changes it so that folder handles are not stored by default. You
need to call `.save_to_idb()` to save it and get the key to restore it.

white-axe committed:
merge: merge branch 'dev' into archive-manager

white-axe committed:
feat(archive manager): start implementing a filesystem view

white-axe committed:
fix(archive manager): fix archive contents not displaying correctly

white-axe committed:
fix(archive manager): fix some UI oddities

white-axe committed:
merge: merge branch 'dev' into archive-manager

white-axe committed:
fix(archive manager): don't persist collapsing header state

We don't want the state of the collapsing headers in the filesystem view
to be saved because we may be opening a different archive with the same
name, or we may be opening a modified version of the same archive. That
would lead to some errors when opening the view.

white-axe committed:
fix(archive manager): depersist the inner subtrees as well

white-axe committed:
style(archive manager): future proofing and use generic IDs

This changes the filesystem view to use the indextree node IDs to
generate the egui IDs for the collapsing headers instead of using the
file paths to generate the egui IDs. This is so that we leave less junk
objects in the memory if the user opens a lot of different archives with
different names.

Also, the response tuple for the collapsing headers is now read using
pattern matching instead of by just indexing into the tuple, which makes
it easier to debug errors if egui decides to change the format of the
response tuple.

white-axe committed:
fix(archive manager): don't load folder contents until partly open

white-axe committed:
feat(archive manager): implement filesystem view file/folder selection

white-axe committed:
perf(archive manager): store `depersisted` in nodes, not hashset

I may be stupid.

white-axe committed:
feat(archive manager): add stripes to filesystem view rows

white-axe committed:
merge: merge branch 'dev' into archive-manager

white-axe committed:
feat(archive manager): implement `SelectedIter`

This iterator allows iterating over the selected entries of a
`FileSystemView` from top to bottom.

white-axe committed:
perf(filesystem): `create_dir` now creates missing parents

This prevents us from having to check if a directory exists before
creating it as well as simplifying the web filesystem implementation and
making the `create_dir` function more consistent with `remove_dir`.

white-axe committed:
feat(archive manager): implement archive extraction

white-axe committed:
perf(archive manager): don't recurse into deselected dirs

white-axe committed:
merge: merge branch 'dev' into archive-manager

white-axe committed:
feat(filesystem): implement file picker for saving files

white-axe committed:
refactor(filesystem): implement `File` for `&mut impl File`

white-axe committed:
refactor(filesystem): use builder pattern to save files

white-axe committed:
refactor(filesystem): make saving files less of a hassle

white-axe committed:
fix: archiver filesystem `create_dir` is now recursive

I changed all of the other filesystems' `create_dir` implementations to
be recursive, so this one needs to be recursive as well.

white-axe committed:
fix(filesystem): fix archiver not being able to create files

white-axe committed:
fix(filesystem): fix edge cases in trie.rs

white-axe committed:
fix(filesystem): fix archiver file creation using wrong magic value

white-axe committed:
feat(archive manager): implement archive creation

white-axe committed:
fix(filesystem): fix more edge cases in `get_dir_prefix`

It would appear I need to do more fuzz testing.

white-axe committed:
fix(filesystem): fix magic calculation for RGSS3A archives

white-axe committed:
perf: improve archive creation speed

The secret sauce is to buffer writes and not flush the archive until
after all files have been written.

white-axe committed:
fix(filesystem): fix trie.rs directory merge algorithm

This is the last edge case, I promise!

white-axe committed:
feat(archive manager): use better sorting in filesystem view

white-axe committed:
fix(archive manager): disable the UI when file/folder picker is active

white-axe committed:
fix(archive manager): use correct file extension when creating archives

white-axe committed:
fix(archive manager): use better error handling

white-axe committed:
feat(archive manager): implement Control and Shift for filesystem view

You can now use the Control and Shift keys like how they normally work
in file managers.

white-axe committed:
perf(archive manager): remove redundant stack from previous commit

white-axe committed:
feat(archive manager): shift-selection now deselects if pivot is deselected

white-axe committed:
feat(archive manager): select "Data" and "Graphics" when creating archives

white-axe committed:
feat(archive manager): load from current project by default

white-axe committed:
chore: clippy

white-axe committed:
fix(archive manager): `is_root` should be set to false for inner calls

melody-rs reviewed (commented):
Seems there's now an error when opening To The Moon, I'm looking into it right now
![292656127-f650c52f-adde-49a7-989d-1207f2c7ee96](https://github.com/Astrabit-ST/Luminol/assets/54482069/77da1efa-0ff4-4a6a-b537-0f0ae5dc4474)

melody-rs commented:
Seems to be erroring here while getting the metadata for `Graphics\Panoramas`
https://github.com/Astrabit-ST/Luminol/blob/a7cd8c4ec0a4cfae7080e5253a6c18cc8219e642/crates/filesystem/src/archiver/filesystem.rs#L648

melody-rs commented:
Looks like it's because of the trie storing `Graphics/Panoramas` and the iterator returning `Graphics\Panoramas`. `Utf8PathBuf.join` uses `\` on Windows and `/` on Linux, which is why I never noticed this in #78 because I was testing this on Linux!

melody-rs commented on crates/components/src/filesystem_view.rs:
```diff
@@ -0,0 +1,718 @@
+                *expanded = header.openness(ui.ctx()) >= 0.2;
+
+                let layout = *ui.layout();
+                header_response = Some(header.show_header(ui, |ui| {
```
It doesn't seem possible to select different files/folders without selecting the folders inbetween?
![luminol_wOq0Hzwpcg](https://github.com/Astrabit-ST/Luminol/assets/54482069/381e60cc-8c7c-4e0d-8ae6-748cac2b7796)

melody-rs replied:
Managed to get Fading Memory running (mostly) off of an rgssad produced by Luminol. I was pleasantly surprised by this, as RPG Maker fails to create archives for games > 128mb!

white-axe replied:
You can select multiple entries by using Ctrl+click on each.

You can select more than one range by using Ctrl+click to select the first endpoint of each range and Ctrl+Shift+click to select the second endpoint.

You can also perform negative selection by first selecting a range, then using Ctrl+click to deselect one endpoint of the negative selection inside of the first selection and Ctrl+Shift+click on the second endpoint of the negative selection.

melody-rs replied:
Could you add a tooltip explaining that?

white-axe replied:
Sure, although these are exactly the semantics from GTK's file picker

melody-rs replied:
Oh. I honestly didn't know that

Looks like it even applies to Windows Explorer 🤦‍♀️

white-axe replied:
I added a tooltip that shows up when hovering over the filesystem view, and it hides once the user selects something

melody-rs reviewed (changes requested):
Still testing archives produced by this, other than the file select and the bug with path separators this looks good to me!

melody-rs commented on crates/ui/src/windows/archive_manager.rs:
```diff
@@ -0,0 +1,560 @@
+        }
+    }
+
+    fn create_archive(
```
Create archive and copy_files should be run on another thread (possibly using `Promise::spawn_blocking`?) to avoid freezing up the ui

white-axe replied:
I implemented `AsyncRead + AsyncWrite + AsyncSeek` for host files by performing IO in another thread on native and asynchronously waiting on the oneshot channels on web, and the archive manager now creates and extracts archives asynchronously

white-axe committed:
fix(filesystem): fix archiver path separator bug on Windows

white-axe committed:
fix(filesystem): fix `FileSystemTrieDirIter` `ExactSizeIterator` implementation

white-axe committed:
fix(filesystem): fix trie.rs `create_dir` edge case

I lied. THIS is the last edge case, for real this time!

This fixes a bug where the trie registers the root directory as a child
of itself if the user calls `create_dir` on the trie with any path and
then calls `create_dir` on the trie a second time with the empty path.

white-axe committed:
feat(filesystem): implement asynchronous IO for host files

Host files now implement `AsyncRead + AsyncWrite + AsyncSeek + Unpin` so
that we can perform IO asynchronously.

Native files do this by spawning a thread and performing IO there (using
the `async-fs` crate).

Web files do this by just asynchronously waiting for command responses
instead of blocking.

white-axe committed:
feat(archive manager): archives are now created asynchronously

white-axe committed:
fix(filesystem): web file futures now clear on error

white-axe committed:
feat(archive manager): archives are now extracted asynchronously

white-axe committed:
feat(archive manager): add progress bars for extraction and creation

white-axe committed:
fix(filesystem): fix trie.rs `remove_dir` not removing entry from parent dir

white-axe committed:
style(filesystem): remove unnecessary `vec!` from `read_file_xor_async`

white-axe committed:
style(filesystem): `get_mut_file` -> `get_file_mut`

white-axe committed:
fix: fix app being drawn every frame on native

white-axe committed:
fix(archive manager): request repaint when progress bar updates

white-axe committed:
style(filesystem): remove poll-promise from filesystem dependencies

white-axe committed:
style(filesystem): fix formatting of imports in archiver/filesystem.rs

white-axe committed:
feat(archive manager): add tooltip explaining how to select files/folders

white-axe committed:
style(filesystem): fix formatting in archiver/filesystem.rs again
